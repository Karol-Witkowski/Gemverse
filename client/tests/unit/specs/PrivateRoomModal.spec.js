import { createLocalVue, mount } from '@vue/test-utils';
import axios from 'axios';
import Vue from 'vue';
import Vuetify from 'vuetify';
import PrivateRoomModal from '@/components/room/PrivateRoomModal.vue';

const localVue = createLocalVue();
const url = 'http://localhost:3000/api/room/verification';
let vuetify;
let wrapper;
const mockRouter = {
  push: jest.fn(),
};
const mockStore = {
  getters: {
    getCurrentRoom: {
      name: 'name',
    },
  },
};
const error = {
  response: {
    data: {
      message: 'error',
    },
  },
};
const response = {
  data: {
    data: {
      slug: 'slug',
    },
  },
};

jest.mock('axios');
vuetify = new Vuetify();

describe('Implementation test for PrivateRoomModal.vue - successful HTTP post', () => {
  beforeEach(() => {
    axios.post.mockResolvedValue(response);

    wrapper = mount(PrivateRoomModal, {
      localVue,
      mocks: {
        $router: mockRouter,
        $store: mockStore,
      },
      vuetify,
      data() {
        return {
          isFormValid: false,
          privateRoomPassword: '',
        };
      },
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
    wrapper.destroy();
  });

  it('Render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Initializes with correct elements', () => {
    // Test buttons initial state
    expect(wrapper.findAll('button').length).toEqual(2);
    expect(wrapper.findAll('button').at(0).text()).toMatch('close');
    expect(wrapper.findAll('button').at(1).text()).toMatch('enter');
    expect(wrapper.findAll('button').at(1).element.disabled).toBeTruthy();

    // Test inputs initial state
    expect(wrapper.findAll('.v-text-field').length).toEqual(1);
    expect(wrapper.findAll('input').at(0).text()).toEqual('');

    // Test validation initial state
    expect(wrapper.findAll('.v-messages').length).toEqual(1);
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('');
    expect(wrapper.vm.isformValid).toBeFalsy();
  });

  it('Set input data correctly', async () => {
    // Check inputs initial state
    expect(wrapper.findAll('.v-text-field').length).toEqual(1);
    expect(wrapper.findAll('input').at(0).text()).toEqual('');

    await wrapper.setData({
      privateRoomPassword: '123456',
    });

    await Vue.nextTick();

    // Check that the password was properly set
    expect(wrapper.findAll('input').at(0).element.value).toEqual('123456');
  });

  it('Fail validation on empty input', () => {
    // Check initial input value
    expect(wrapper.findAll('input').at(0).text()).toEqual('');

    // Check validation state
    expect(wrapper.vm.isFormValid).toBeFalsy();

    // Check that correct input message is displayed
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('');
  });

  it('Enables enter room button on validation pass', async () => {
    await wrapper.setData({
      privateRoomPassword: '123456',
    });

    await Vue.nextTick();

    // Check validation state
    expect(wrapper.vm.isFormValid).toBeTruthy();

    // Check that correct input message are displayed
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('');

    // Check that the enter button is active
    expect(wrapper.findAll('button').at(1).element.disabled).toBeFalsy();
  });

  it('Should sends post request with correct data on submit', async () => {
    await wrapper.setData({
      privateRoomPassword: '123456',
    });

    await Vue.nextTick();

    wrapper.vm.passwordValidation();

    // Check if post was called
    expect(axios.post).toHaveBeenCalled();

    // Check if post was called once
    expect(axios.post).toHaveReturnedTimes(1);

    // Check if post was called with correct data
    expect(axios.post).toHaveBeenCalledWith(
      url,
      {
        name: 'name',
        password: '123456',
      },
    );
  });

  it('Push user to room', async () => {
    wrapper.vm.join();

    // Check if router push was called
    expect(mockRouter.push).toHaveBeenCalled();

    // Check if router push was called one time
    expect(mockRouter.push).toHaveBeenCalledTimes(1);

    // Check if router push was called with correct object
    expect(mockRouter.push).toHaveBeenCalledWith({
      name: 'Room',
      params: { slug: undefined },
    });
  });

  it('Should emit close dialog event', async () => {
    await wrapper.setData({
      privateRoomPassword: '123456',
    });

    await wrapper.vm.closeModal();

    expect(wrapper.emitted().closeModal).toBeTruthy();
    expect(wrapper.emitted().closeModal.length).toBe(1);
  });
});

describe('Implementation test for PrivateRoomModal.vue - failed HTTP post', () => {
  beforeEach(() => {
    axios.post.mockRejectedValue(error);

    wrapper = mount(PrivateRoomModal, {
      localVue,
      mocks: {
        $store: mockStore,
      },
      vuetify,
      data() {
        return {
          error: '',
          isFormValid: false,
          privateRoomPassword: '',
        };
      },
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
    wrapper.destroy();
  });

  it('Should display errors', async () => {
    await wrapper.setData({
      privateRoomPassword: '123456',
    });

    await Vue.nextTick();

    await wrapper.vm.passwordValidation();

    await Vue.nextTick();

    expect(wrapper.findAll('.v-messages').length).toEqual(1);
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('error');
    expect(wrapper.vm.isformValid).toBeFalsy();
  });
});

describe('Behavioral test for PrivateRoomModal.vue - successful HTTP post', () => {
  beforeEach(() => {
    axios.post.mockResolvedValue(response);

    wrapper = mount(PrivateRoomModal, {
      localVue,
      mocks: {
        $router: mockRouter,
        $store: mockStore,
      },
      vuetify,
      data() {
        return {
          isFormValid: false,
          privateRoomPassword: '',
        };
      },
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
    wrapper.destroy();
  });

  it('Should not sends post request when input is empty', async () => {
    await wrapper.findAll('button').at(1).trigger('click');

    expect(axios.post).not.toHaveBeenCalled();
  });

  it('Set input data correctly', async () => {
    await wrapper.findAll('input').at(0).setValue('123456');

    await Vue.nextTick();

    // Check that the password was properly set
    expect(wrapper.findAll('input').at(0).element.value).toEqual('123456');
  });

  it('Should send post request on submit', async () => {
    await wrapper.findAll('input').at(0).setValue('123456');

    await Vue.nextTick();

    wrapper.findAll('button').at(1).trigger('click');

    // Check if post was called
    expect(axios.post).toHaveBeenCalled();

    // Check if post was called once
    expect(axios.post).toHaveReturnedTimes(1);

    // Check if post was called with correct data
    expect(axios.post).toHaveBeenCalledWith(
      url,
      {
        name: 'name',
        password: '123456',
      },
    );
  });
});

describe('Behavioral test for PrivateRoomModal.vue - failed HTTP post', () => {
  beforeEach(() => {
    axios.post.mockRejectedValue(error);

    wrapper = mount(PrivateRoomModal, {
      localVue,
      mocks: {
        $store: mockStore,
      },
      vuetify,
      data() {
        return {
          error: '',
          isFormValid: false,
          privateRoomPassword: '',
        };
      },
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
    wrapper.destroy();
  });

  it('Should reset errors on keyup event', async () => {
    await wrapper.setData({
      privateRoomPassword: '123456',
    });

    await Vue.nextTick();

    await wrapper.findAll('button').at(1).trigger('click');

    await Vue.nextTick();

    // Check if error is visible
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('error');

    await wrapper.findAll('input').trigger('keyup');

    // Check if keyup event reset error
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('');
  });

  it('Should reset error on close event', async () => {
    await wrapper.setData({
      privateRoomPassword: '123456',
    });

    await Vue.nextTick();

    await wrapper.findAll('button').at(1).trigger('click');

    await Vue.nextTick();

    // Check if error is visible
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('error');

    await wrapper.findAll('button').at(0).trigger('click');

    // Check if closing modal reset error
    expect(wrapper.findAll('.v-messages').at(0).text()).toEqual('');
  });
});
