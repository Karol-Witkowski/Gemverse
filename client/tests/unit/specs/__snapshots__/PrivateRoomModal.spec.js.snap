// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Implementation test for PrivateRoomModal.vue - successful HTTP post Render correctly 1`] = `
<div class="v-card v-sheet theme--light" cols="12">
  <div class="v-card__title header headline grey--text text--darken-2">
    Type password to proceed
  </div>
  <div class="v-card__text">
    <form novalidate="novalidate" type="submit" onsubmit="return false" class="v-form">
      <div class="col col-12">
        <div class="v-input theme--light v-text-field">
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-4" class="v-label theme--light" style="left: 0px; position: absolute;">Room password</label><input name="password" required="required" id="input-4" type="password"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
  <div class="v-card__actions pb-4"><button type="button" class="v-btn v-btn--outlined v-btn--text theme--light v-size--default primary--text" name="close"><span class="v-btn__content">
      close
    </span></button>
    <div class="spacer"></div> <button type="submit" disabled="disabled" class="v-btn v-btn--disabled v-btn--outlined v-btn--text theme--light v-size--default" name="enter"><span class="v-btn__content">
      enter
    </span></button>
  </div>
</div>
`;
